standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java

```diff
@@ -20,71 +20,646 @@
+          LOG.warn("Unable to instantiate custom database product. Reverting to default", e);
+        }
+      }
+      else {
+        LOG.warn("metastore.use.custom.database.product was set, " +
+                 "but metastore.custom.database.product.classname was not. Reverting to default");
+      }
+    }
+  }
+
+  /**
+   * Private constructor for singleton class
+   * @param id
+   */
+  private DatabaseProduct() {}
+  
+  public static final String DERBY_NAME = "derby";
+  public static final String SQL_SERVER_NAME = "microsoft sql server";
+  public static final String MYSQL_NAME = "mysql";
+  public static final String POSTGRESQL_NAME = "postgresql";
+  public static final String ORACLE_NAME = "oracle";
+  public static final String OTHER_NAME = "other";
+  
   /**
    * Determine the database product type
    * @param productName string to defer database connection
    * @return database product type
    */
-  public static DatabaseProduct determineDatabaseProduct(String productName) throws SQLException {
+  public static DatabaseProduct determineDatabaseProduct(String productName) {
+    DbType dbt;
+
     if (productName == null) {
-      return OTHER;
+      productName = OTHER_NAME;
     }
+
     productName = productName.toLowerCase();
-    if (productName.contains("derby")) {
-      return DERBY;
-    } else if (productName.contains("microsoft sql server")) {
-      return SQLSERVER;
-    } else if (productName.contains("mysql")) {
-      return MYSQL;
-    } else if (productName.contains("oracle")) {
-      return ORACLE;
-    } else if (productName.contains("postgresql")) {
-      return POSTGRES;
+
+    if (productName.contains(DERBY_NAME)) {
+      dbt = DbType.DERBY;
+    } else if (productName.contains(SQL_SERVER_NAME)) {
+      dbt = DbType.SQLSERVER;
+    } else if (productName.contains(MYSQL_NAME)) {
+      dbt = DbType.MYSQL;
+    } else if (productName.contains(ORACLE_NAME)) {
+      dbt = DbType.ORACLE;
+    } else if (productName.contains(POSTGRESQL_NAME)) {
+      dbt = DbType.POSTGRES;
     } else {
-      return OTHER;
+      dbt = DbType.OTHER;
     }
+
+    // This method may be invoked by concurrent connections
+    synchronized (DatabaseProduct.class) {
+      if (theDatabaseProduct == null) {
+        theDatabaseProduct = new DatabaseProduct();
+      }
+  
+      theDatabaseProduct.dbType = dbt;
+    }
+    return theDatabaseProduct;
+  }
+
+  public final boolean isDERBY() {
+    return dbType == DbType.DERBY;
+  }
+
+  public final boolean isMYSQL() {
+    return dbType == DbType.MYSQL;
   }
 
-  public static boolean isDeadlock(DatabaseProduct dbProduct, SQLException e) {
+  public final boolean isORACLE() {
+    return dbType == DbType.ORACLE;
+  }
+
+  public final boolean isSQLSERVER() {
+    return dbType == DbType.SQLSERVER;
+  }
+
+  public final boolean isPOSTGRES() {
+    return dbType == DbType.POSTGRES;
+  }
+
+  public final boolean isEXTERNAL() {
+    return dbType == DbType.EXTERNAL;
```

Comment:
Its getting a bit confusing with the usage of external, other, custom. Can this be avoided? If not, we should document it somewhere in this class.

Reply:
Yeah. From what I see in the codebase, OTHER seems to really mean UNDEFINED. I've changed the constant name accordingly. I believe this addresses the possible confusion. Let me know otherwise.